addon/services/prefetch.js

```diff
@@ -25,14 +25,16 @@ if (gte('3.6.0')) {
             let changeSet = createPrefetchChangeSet(privateRouter, transition);
             if (changeSet.shouldCall) {
               for (let i = 0; i < changeSet.for.length; i++) {
-                let { route, fullParams } = changeSet.for[i];
-                if (seenRoutes.has(route)) continue;
-
-                  route._prefetched = new RSVP.Promise(r => {
-                    return r(route.prefetch(fullParams, transition))
+                // normalize route to always be a promise since lazy loaded
+                // routes are promises, while preloaded routes are not. We want
+                // to treat them both the same.
+                RSVP.Promise.resolve(changeSet.for[i]).then(({ route, fullParams }) => {
```

Comment:
It does seem odd that I am wrapping `changeSet.for[i]` in a promise when it never actually could be a promise. I had some concern, however, that we could end up using the last `fullParams` object inside the `then` handler for doing prefetch when we did not mean to if we don't make sure to capture both the route and the fullParams at the current iteration to pass into the `then` callback. My intuition about this may be wrong and I may just not be thinking through it clearly, but it seemed like this could happen. Or there could be a completely different way to handle the situation.